ONSdigital/dp-topic-api
# Add AuthN & AuthZ to topics/<id> endpoint

### What

1 - Add in Authentication and Authorisation code from dp-dataset-api
2 - Add test code to exercise Public and Private endpoints
3 - Add extra bits of code for future use (there are !!! markers in code to identify future changes)
4 - Use the topics collection index “id”
5 - Restructure code to follow pattern in dp-dataset-api to make it easier to add future features

### How to review

Inspect code changes.

If you want to test, do:

To test, Follow instructions in ‘Section 3’ of document dp-topics/db/db-tests.md

[ and note different script being used (db-init-4.js):
To empty and populate the database, follow steps (A. to C.):
(you need to have mongodb 3.4+ installed and running)
A. In dp-topics-api directory, cd into db directory and enter command: mongo db-init-4.js
B. Frig the App to run locally by replacing the following variables in main.go as such:
    BuildTime string = "1601119818"
    GitCommit string = "6584b786caac36b6214ffe04bf62f058d4021538"
    Version string = "v0.1.0"

And run the App.
C. In browser, enter url: http://localhost:25300/topics/1
   to see new: current, next structure
]

To test in publishing:
    run up the publishing stack with zebedee, etc running.
    Do: echo $zebedee_root
and this gave me:
    /Users/rhysdavies/Documents/website/zebedee-content/generated
and in there, drill down into:
    zebedee/services
to find a .json file whose name is 64 hex characters.
Copy the 64 hex character name to use in curl command below : replacing <64 HEX name>

The Environment variable: ENABLE_PRIVATE_ENDPOINTS needs to be true

Run the dp-topic-api (with adjustments to main variables for it to run locally), and in a terminal enter:
curl -X GET localhost:25300/topics/1 -H "Authorization: Bearer <64 HEX name>" |jq -C
curl -X GET localhost:25300/topics/6 -H "Authorization: Bearer <64 HEX name>" |jq -C

Use postman to see 401/404 error responses for topic requests

To test in web:
The Environment variable: ENABLE_PRIVATE_ENDPOINTS needs to be false

Re-run dp-topic-api and repeat the curl and postman tests without Authorization.

### Who can review

Anyone
